### DISCUSSION TITLE
Sharing components with actions between pages

### DISCUSSION THREAD

**Discussion by tyteen4a03** [remix-run/remix#5295] [Q&A] [answered]
This is a pain point I have been running into a lot. 

I have a widget `<LookupThingWidget>` that calls an action (say `POST /lookup/thing`) and depending on the results it returns, does something on the client (like showing validation errors, redirecting, or emitting `toast()` notifications on error/success, something that Remix also doesn't do well, but that's another topic for another day). I want to be able to reuse it into any page that I want (`/a/b/c`, `/d/e`, `/f`, you get the idea). This breaks Remix's MPA model, and to solve this I have two options:

1. Change `<LookupThingWidget>` to `POST` to the page it is in, and make _all_ actions for _all_ pages that expect the widget understand the `<LookupThingWidget>` subform - this is very costly and increases friction as the amount of widgets I want to reuse grows
2. Use react-query's `useMutation` to hoist the call above the page hierarchy - this is the method I am currently using, however everything Remix does goes against this, so I'm finding myself reinventing a lot of wheels, especially with form validation error handling (I'm using `remix-validated-form`)

What is the Remix way of solving this issue that does not involve solution 1, and offer better ergonomics than 2?

**Comment by machour** [comment, marked as answer]
I think this is a perfect case for a full stack component: https://www.epicweb.dev/full-stack-components

**Reply by tyteen4a03** [reply]
Wow - how is this not in the official docs?

**Reply by machour** [reply]
Glad you liked it! I'll see if we can add a mention and a link to that in the docs 😃
